docs: how to install without generated resources (#713)
hashicorp/boundary: website/content/docs/installing/no-gen-resources.mdx

```

</Tab>
<Tab heading="Terraform">

```hcl
resource "boundary_account" "myuser" {
  name           = "myuser"
  description    = "User account for my user"
  type           = "password"
  login_name     = "myuser" 
  password       = "foofoofoo"
  auth_method_id = boundary_auth_method.password.id
}
```

</Tab>
</Tabs>

### Create a User

Create a user and associate the user with the login account created in the previous step.
This user will also be the principal in the role we create in the following step.

<Tabs>
<Tab heading="CLI">

```bash
$ boundary users create -scope-id <org_scope_id> \
  -recovery-config /tmp/recovery.hcl \
  -name "myuser" \
  -description "My user!"

$ boundary users add-accounts -id <myuser_user_id> -account <myuser_account_id>
```

</Tab>
<Tab heading="Terraform">

```hcl
resource "boundary_user" "myuser" {
  name        = "myuser"
  description = "My user!"
  account_ids = [boundary_account.myuser.id]
  scope_id    = boundary_scope.org.id
}
```

</Tab>
</Tabs>

### Create Roles to Manage Scopes

The following describes the four baseline roles you'll need to create to manage resources within the org and project 

> scopes created above. These roles are similar to the roles created for you in dev mode but declaring them explicitly

Reviewer: ```suggestion
scopes created above. These roles are similar to the roles created for you in if generation had not been skipped during `boundary database init`, but declaring them explicitly
```

  -recovery-config /tmp/recovery.hcl \
  -login-name "myuser" \
  -password "foofoofoo" \
  -auth-method-id <auth_method_id_from_last_step"
```

</Tab>
<Tab heading="Terraform">

```hcl
resource "boundary_account" "myuser" {
  name           = "myuser"
  description    = "User account for my user"
  type           = "password"
  login_name     = "myuser" 
  password       = "foofoofoo"
  auth_method_id = boundary_auth_method.password.id
}
```

</Tab>
</Tabs>

#### Create a User

Create a user and associate the user with the login account created in the previous step.
This user will also be the principal in the role we create in the following step.

<Tabs>
<Tab heading="CLI">

```bash
$ boundary users create -scope-id <org_scope_id> \
  -recovery-config /tmp/recovery.hcl \
  -name "myuser" \
  -description "My user!"

$ boundary users add-accounts -id <myuser_user_id> -account <myuser_account_id>
```

</Tab>
<Tab heading="Terraform">

```hcl
resource "boundary_user" "myuser" {
  name        = "myuser"
  description = "My user!"
  account_ids = [boundary_account.myuser.id]
  scope_id    = boundary_scope.org.id
}
```

</Tab>
</Tabs>

#### Create Roles to Manage Scopes

> 

Reviewer: It might be worth some text here explaining to the user that these four roles will end up granting the same access as if role autogeneration had been enabled, but with Terraform in full control?

The follow example creates 4 roles:
1. To allow anonymouns (unauthenticated) users the ability to list scopes and auth methods in the global scope.
1. To allow anonymouns (unauthenticated) users the ability to list scopes and auth methods in the orgnaization scope.
1. To allow `myuser` user administration grants at the org scope.
1. To allow `myuser` user administration grants at the project scope.

1. Create anonymous listing role for global scope

<Tabs>
<Tab heading="CLI">

Assumes recovery key export from above steps is still set:

```bash
# Create global anonymous listing role
$ boundary roles create -name 'global_anon_listing' \
  -recovery-config /tmp/recovery.hcl \
  -scope-id 'global'

> $ boundary add-grants -id <global_anon_listing_id> \

Reviewer: you can specify `-grant` multiple times so this can be one call instead of three.

  -recovery-config /tmp/recovery.hcl \
  -grant 'id=*;type=auth-method;actions=list,authenticate'

$ boundary add-grants -id <global_anon_listing_id> \
  -recovery-config /tmp/recovery.hcl \
  -grant 'type=scope;actions=list'

$ boundary add-grants -id <global_anon_listing_id> \
  -recovery-config /tmp/recovery.hcl \

>   -grant 'id=*;actions=read,change-password'

Reviewer: I think you meant for this to be `'id={{account.id}};actions=read,change-password'`

$ boundary add-principals -id <global_anon_listing_id> \
  -recovery-config /tmp/recovery.hcl \
  -principal 'u_anon' 
```

</Tab>
<Tab heading="Terraform">

```hcl
resource "boundary_role" "global_anon_listing" {
  scope_id = "global"
  grant_strings = [
    "id=*;type=auth-method;actions=list,authenticate",
    "type=scope;actions=list",

>     "id=*;actions=read,change-password"

Reviewer: Same here

  ]
  principal_ids = ["u_anon"]
}
```

</Tab>
</Tabs>

2. Create anonymous listing role for org scope

<Tabs>
<Tab heading="CLI">

Assumes recovery key export from above steps is still set:

```bash

> $ boundary roles create -name 'global_anon_listing' \

Reviewer: Did you mean for global_anon_listing here and below to be something like "org_anon_listing"?

resources. 

To authenticate to Boundary using the recovery KMS workflow:

<Tabs>
<Tab heading="CLI">

To use the recovery workflow on the CLI, you must pass the `-recovery-config` flag or set the environment
variable for `BOUNDARY_RECOVERY_CONFIG` for every command ran. Authentication takes place for every command
ran when using the recovery workflow, there is no `boundary authenticate` step:

```bash
$ export BOUNDARY_RECOVERY_CONFIG=<<EOT
kms "aead" {
	purpose   = "recovery"
	aead_type = "aes-gcm"
	key       = "8fZBjCUfN0TzjEGLQldGY4+iE9AkOvCfjh7+p0GtRBQ="
	key_id    = "global_recovery"
}
EOT

$ boundary users create ... 
...
```

</Tab>
<Tab heading="Terraform">

To configure your provider to use the recovery KMS workflow, provide the KMS block as the value for
`recovery_kms_hcl`:

```hcl
provider "boundary" {
  addr             = 'https://boundary.mycorp.com:9200' 
  recovery_kms_hcl = <<EOT
kms "aead" {
	purpose   = "recovery"
	aead_type = "aes-gcm"
	key       = "8fZBjCUfN0TzjEGLQldGY4+iE9AkOvCfjh7+p0GtRBQ="
	key_id    = "global_recovery"
}
EOT
}
```

</Tab>
</Tabs>

## Initialize the Database

> Before you can start Boundary, the database must be initialized. To do this, run:

Reviewer: This doesn't _actually_ initialize the database. Maybe say:

```suggestion
Before you can start Boundary, the database must be initialized. It's useful to look at the help output for the init command:
```

```
$ boundary database init -h
...
```

From this command, you can see the flags available to skip the creation of auto-generated resources. These
resources are useful when running in development environments but in a non development environment may not

> be desired. You can think of them as Boundary's default username and password, and if used in non dev workflows

Reviewer: I don't really understand this sentence. What about the generated resources undermines Boundary's security model?

Author: For example, if you don't change the default login account's password, it's hardcoded.

Reviewer: Only in dev mode. Everything is fully generated when doing `boundary database init` and you _cannot_ pick the values.

should be overidden or skipped to avoid undermining Boundary's security model.

To initialize the Boundary database without generated resources:

```
$ boundary database init -skip-initial-login-role -config /etc/boundary.hcl
```

> When you start Boundary, you will effectively have a blank sheet to work against. The initial migrations in the database have been run (note that this includes creating special users like `u_anon` and the `global` scope) and the internal keyrings have been initialized.. From here, it's required that

Reviewer: ```suggestion
When you start Boundary, you will effectively have a blank sheet to work against. The initial migrations in the database have been run (note that this includes creating special users like `u_anon` and the `global` scope) and the internal keyrings have been initialized. From here, it's required that
```

you use the KMS recovery workflow described above to create at a minimum an auth method, a user, an account, and a 

> role with sufficient grants. Otherwise, you need to continue to use the recovery workflow for management. t's important 

Reviewer: ```suggestion
role with sufficient grants. Otherwise, you need to continue to use the recovery workflow for management. It's important 
```

the recovery KMS workflow. In this example, we're going to make an admin user for the global and project level
scopes we create. This will allow our user to configure targets within those scopes and manage them.

#### Create Org and Project Scopes

In this example, we're going to create an org and project scope and skip creating an administrator and admin role 
for each scope. We're going to specify a role for managing these scopes by selected users in a later step.

<Tabs>
<Tab heading="CLI">

```bash
$ export BOUNDARY_RECOVERY_CONFIG=<<EOT
kms "aead" {
	purpose   = "recovery"
	aead_type = "aes-gcm"
	key       = "8fZBjCUfN0TzjEGLQldGY4+iE9AkOvCfjh7+p0GtRBQ="
	key_id    = "global_recovery"
}
EOT

$ boundary scopes create -name 'org' -scope-id 'global' \
  -skip-admin-role-creation \
  -skip-default-role-creation
<truncated>

$ boundary scopes create -name 'project' -scope-id <org_scope_id_from_last_step> \
  -skip-admin-role-creation \
  -skip-default-role-creation
<truncated>
```

</Tab>
<Tab heading="Terraform">

```hcl
resource "boundary_scope" "org" {
  scope_id    = "global"
  name        = "organization"
  description = "Organization scope"

>   auto_create_admin_role   = false 

Reviewer: These are the defaults, you don't need to specify them. Not doing so makes it more clear in the other examples when you _are_ specifying them that that is not the default behavior.

Author: I went back and forth on this, I wanted to explicitly call them out since the logic on the CLI is other way around: in TF, it's the default but on the CLI, you have to explicitly call the flag. If someone was looking at the TF and CLI examples side by side, it would be more clear.

Reviewer: Fair enough

---
layout: docs

> page_title: Installing Boundary

Reviewer: ```suggestion
page_title: Installing Boundary Without Generated Resources
```

sidebar_title: Without Generated Resources 
description: |-
  How to install boundary without generated resources.
---

# Installing Boundary Without Generated Resources 

> Initializing Boundary without Generated resources for the first time can be a little confusing. How do you 

Reviewer: ```suggestion
Initializing Boundary without generated resources for the first time can be a little confusing. How do you 
```

should use your cloud provider's KMS such as [AWS KMS](https://www.boundaryproject.io/docs/configuration/kms/awskms) to manage the keys Boundary
users to encrypted sensitive information.

The KMS block we're focused on is the `recovery` block. This block specifies the key used to "recover" Boundary
but you can also use it to authenticate to Boundary to manage it as a "global" super user. This allows
you to authenticate from the CLI or from Terraform in order to manage Boundary without any generated 
resources. 

To authenticate to Boundary using the recovery KMS workflow:

<Tabs>
<Tab heading="CLI">

To use the recovery workflow on the CLI, you must pass the `-recovery-config` flag or set the environment
variable for `BOUNDARY_RECOVERY_CONFIG` for every command ran. Authentication takes place for every command
ran when using the recovery workflow, there is no `boundary authenticate` step:

```bash
$ export BOUNDARY_RECOVERY_CONFIG=<<EOT
kms "aead" {
	purpose   = "recovery"
	aead_type = "aes-gcm"
	key       = "8fZBjCUfN0TzjEGLQldGY4+iE9AkOvCfjh7+p0GtRBQ="
	key_id    = "global_recovery"
}
EOT

> $ boundary connect ssh -target-id ttcp_873964g581

Reviewer: You specifically _cannot_ use the recovery workflow to authorize a session! :-)

...
```

</Tab>
<Tab heading="Terraform">

To configure your provider to use the recovery KMS workflow, provide the KMS block as the value for
`recovery_kms_hcl`:

```hcl
provider "boundary" {
  addr             = 'https://boundary.mycorp.com:9200' 
  recovery_kms_hcl = <<EOT
kms "aead" {
	purpose   = "recovery"
	aead_type = "aes-gcm"
	key       = "8fZBjCUfN0TzjEGLQldGY4+iE9AkOvCfjh7+p0GtRBQ="
	key_id    = "global_recovery"
}
EOT
}
```

</Tab>
</Tabs>

## Initialize the Database

Before you can start Boundary, the database must be initialized. To do this, run:

```
$ boundary database init -h
...
```

From this command, you can see the flags available to skip the creation of auto-generated resources. These
resources are useful when running in development environments but in a non development environment may not
be desired. You can think of them as Boundary's default username and password, and if used in non dev workflows
should be overidden or skipped to avoid undermining Boundary's security model.

To initialize the Boundary database without generated resources:

```
$ boundary database init \

>   -skip-auth-method-creation \

Reviewer: All you need is `-skip-initial-login-role`. If you look at the help output for that command, it does generation in-order and skipping anything higher up the ordering skips everything below it. For completeness you could leave all of them in there, but then you're missing `-skip-initial-login-role` :-)

Author: 👍

  -skip-host-resources-creation \
  -skip-scopes-creation \
  -skip-target-creation \ 
  -config /etc/boundary.hcl
```

> When you start Boundary, you will effectively have a blank sheet to work against. From here, it's required that

Reviewer: ```suggestion
When you start Boundary, you will effectively have a blank sheet to work against. The initial migrations in the database have been run (note that this includes creating special users like `u_anon` and the `global` scope) and the internal keyrings have been initialized.. From here, it's required that
```

you use the KMS recovery workflow described above to create at a minimum an auth method, a user, an account, and a 
role with sufficient grants. Otherwise, you need to continue to use the recovery workflow for management. Though 

> continuing with the recovery workflow is not discouraged it's important to realize that this is effectively a 

Reviewer: I think saying it's not discouraged is a little _too_ encouraging. Maybe just state the warning and that in some circumstances it may be desirable to continue to use the workflow, but it must be done with caution.
